plugins/cryptolib/cryptolib.cpp

```diff
@@ -0,0 +1,635 @@
+
+    //temporary buffer for result
+    MemoryBuffer out;
+    aesEncrypt(out, lenInputdata, inputdata, strlen(key), key, (const char *)iv);
+
+    //build result DATA buffer in the form:
+    // (size32_t)EVP_MAX_IV_LENGTH, IV, (size32_t)LenPlainText excluding NULL, (size32_t)LenCipher, Cipher
+    symmSerialize(__result, __lenResult, (const char *)iv, sizeof(iv), lenInputdata, out.length(), (const void *)out.bufferBase());
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clSymmetricDecrypt(size32_t & __lenResult,void * & __result,
+                                                const char * algorithm,
+                                                const char * key,
+                                                size32_t lenEncrypteddata,const void * encrypteddata)
+{
+    verifySymmetricAlgorithm(algorithm, strlen(key));
+
+    //Decompose DATA buffer
+
+    StringBuffer sbIV;
+    StringBuffer sbCipher;
+    size32_t     lenPlainText;
+    symmDeserialize(encrypteddata, sbIV, sbCipher, &lenPlainText);
+
+    __result = (char *)CTXMALLOC(parentCtx, lenPlainText);
+    __lenResult = lenPlainText;
+
+    MemoryBuffer decrypted;
+    size32_t len = aesDecrypt(decrypted, sbCipher.length(), sbCipher.str(), strlen(key), key, sbIV.str());
+    memcpy(__result, decrypted.toByteArray(), __lenResult);
+}
+
+
+
+//-------------------------------------------------------------------------------------------------------------------------------------------
+// HASH SUPPORT
+//-------------------------------------------------------------------------------------------------------------------------------------------
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clInstalledHashAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    loadDigests();
+#endif
+    OBJ_NAME_do_all_sorted(OBJ_NAME_TYPE_MD_METH, my_clAlgorithmsCallback, (void*)&algorithms);
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clSupportedHashAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+    algorithms.appendList("SHA1,SHA224,SHA256,SHA384,SHA512", ",");//make sure these match algorithms in clHash()
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+void throwHashError(const char * str)
+{
+    unsigned long err = ERR_get_error();
+    char errStr[1024];
+    ERR_error_string_n(err, errStr, sizeof(errStr));
+    throw MakeStringException(-1, "%s : ERROR %ld (0x%lX) : %s", str, err, err, errStr);
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clHash(size32_t & __lenResult, void * & __result,
+                                         const char * algorithm,
+                                         size32_t lenInputData, const void * inputData)
+{
+    //Make sure each algorithm is included in clHashAlgorithms()
```

Comment:
Does this (and all the other encrption/decryption functions need to call loadDigests()?

Reply:
This question hasn't been answered.

Reply:
Since this code is dependent on cryptohelper, and cryptohelper calls the function OpenSSL_add_all_algorithms at initialization time, why is any call to loadDigests required?
A comment to that effect should be sufficient.